sdk/appconfiguration/azure-spring-cloud-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/implementation/AppConfigurationPropertySourceLocator.java

```diff
@@ -0,0 +1,301 @@
+
+                        // Setting new ETag values for Watch
+                        List<ConfigurationSetting> watchKeysSettings = getWatchKeys(client,
+                            configStore.getMonitoring().getTriggers());
+                        getFeatureFlagWatchKeys(client, configStore, newState);
+
+                        newState.setState(configStore.getEndpoint(), watchKeysSettings,
+                            configStore.getMonitoring().getRefreshInterval());
+                        newState.setLoadState(configStore.getEndpoint(), true);
+                        generatedPropertySources = true;
+                    } catch (AppConfigurationStatusException e) {
+                        reloadFailed = true;
+                        clientFactory.backoffClientClient(configStore.getEndpoint(), client.getEndpoint());
+                    } catch (Exception e) {
+                        newState = failedToGeneratePropertySource(configStore, newState, e);
+
+                        // Not a retiable error
+                        break;
+                    }
+                    if (generatedPropertySources) {
+                        break;
+                    }
+                }
+
+                if (generatedPropertySources) {
+                    // Updating list of propertySources
+                    sourceList.forEach(composite::addPropertySource);
+                } else if (!STARTUP.get() || (configStore.isFailFast() && STARTUP.get())) {
+                    String message = "Failed to generate property sources for " + configStore.getEndpoint();
+
+                    // Refresh failed for a config store ending attempt
+                    failedToGeneratePropertySource(configStore, newState, new RuntimeException(message));
+                }
+
+            } else if (!configStore.isEnabled() && loadNewPropertySources) {
+                LOGGER.info("Not loading configurations from {} as it is not enabled.", configStore.getEndpoint());
+            } else {
+                LOGGER.warn("Not loading configurations from {} as it failed on startup.", configStore.getEndpoint());
+            }
+        }
+
+        StateHolder.updateState(newState);
+        STARTUP.set(false);
+
+        return composite;
+    }
+
+    private List<ConfigurationSetting> getWatchKeys(AppConfigurationReplicaClient client,
+        List<AppConfigurationStoreTrigger> triggers) {
+        List<ConfigurationSetting> watchKeysSettings = new ArrayList<>();
+        for (AppConfigurationStoreTrigger trigger : triggers) {
+            ConfigurationSetting watchKey = client.getWatchKey(trigger.getKey(), trigger.getLabel());
+            if (watchKey != null) {
+                watchKeysSettings.add(watchKey);
+            } else {
+                watchKeysSettings
+                    .add(new ConfigurationSetting().setKey(trigger.getKey())
+                        .setLabel(trigger.getLabel()));
+            }
+        }
```

Comment:
Could you help explain this code? Why either read the `ConfigurationSetting` from the server or construct a new one from the trigger?

Reply:
The use case for the second one is if the watch key wan't in the config store. This creates a fake watch key, so if it ever exists it will be already recorded and the etag will be different as it will be different than empty string.